cxx-qt-lib/build.rs

```diff
@@ -3,138 +3,77 @@
-#[derive(Serialize, Deserialize)]
-struct GeneratedType {
-    header: String,
-    name: String,
-    source: String,
-}
-
-/// Generate a CXX header, source, name for a given Rust file
-fn gen_cxx_sources(folder: &str, file_stem: &str) -> GeneratedType {
-    // Read the rust source files
-    let path = format!("{}/{}.rs", folder, file_stem);
-    println!("cargo:rerun-if-changed={}", path);
-    let content = std::fs::read_to_string(path).expect("Could not read Rust file");
-    let file = syn::parse_file(&content).unwrap();
-
-    // Generate the CXX header and cc for the file
-    let opt = cxx_gen::Opt::default();
-    let generated = cxx_gen::generate_header_and_cc(file.into_token_stream(), &opt)
-        .expect("Could not generate C++ from Rust file");
-
-    GeneratedType {
-        header: String::from_utf8(generated.header).unwrap(),
-        name: format!("{}_cxx", file_stem),
-        source: String::from_utf8(generated.implementation).unwrap(),
-    }
-}
-
-/// Write generates types to a given file as JSON
-fn write_cxx_sources(gen: &Vec<GeneratedType>, path: &str) {
-    let file = std::fs::File::create(path).expect("Could not create generated file");
-    serde_json::to_writer(file, &gen).unwrap();
-}
-
-fn create_and_write_file(path: &impl AsRef<std::path::Path>, file_contents: &str) {
-    let path = path.as_ref();
-    File::create(&path)
-        .unwrap_or_else(|_| panic!("Could not create file {}", path.display()))
-        .write_all(file_contents.as_bytes())
-        .unwrap_or_else(|_| panic!("Could not write file {}", path.display()));
-}
+use std::env;
 
 fn main() {
-    // Read the cargo folder and out folder
-    let mut dir_manifest = std::env::var("CARGO_MANIFEST_DIR").expect("Could not get manifest dir");
-    if cfg!(windows) {
-        dir_manifest = dir_manifest.replace('\\', "/");
+    let qt_modules = vec!["Core", "Gui"]
+        .iter()
+        .map(|m| String::from(*m))
+        .collect();
+    let qtbuild = qt_build::QtBuild::new(qt_modules).expect("Could not find Qt installation");
+    qtbuild.cargo_link_libraries();
+
+    // Copy qt_types.h so C++ build systems can #include it.
+    // By design, CARGO_TARGET_DIR is not set by cargo when running build scripts.
+    // Copying the header is only needed for making the header available to a C++
+    // build system, in which case CARGO_TARGET_DIR will be set by
+    // the C++ build system.
+    println!("cargo:rerun-if-changed=include/qt_types.h");
+    println!("cargo:rerun-if-env-changed=CARGO_TARGET_DIR");
+    if let Ok(target_dir) = env::var("CARGO_TARGET_DIR") {
+        let manifest_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+        std::fs::create_dir_all(&format!("{}/cxxbridge/cxx-qt-lib/include", target_dir)).unwrap();
+        // FIXME: Horrible hack around this sometimes failing because
+        // Windows doesn't allow multiple processes to access a file by default.
+        loop {
```

Comment:
assuming this does work, can we only compile the `loop` for Windows as well ? or do we rely on the `break`, `continue`, and panic. And should there be a sleep to avoid burning CPU ?

Reply:
I added a 100ms sleep, but I don't think compiling the loop only for Windows would be worth the trouble.